src/common/conn/server.rs

```diff
@@ -310,161 +410,122 @@ where
-
-        let app_client = AppClientBuilder::new(grpc_client, self.app_config.clone())
-            .build()
-            .unwrap();
-
-        WebRtcListener::new(self.app_config.get_ip(), app_client, cert, dtls, cloned)
+    pub fn new(cert: Rc<CC>, dtls: D, exec: Executor<'a>) -> Self {
+        Self { dtls, cert, exec }
     }
 }
-
-struct WebRtcConnector<'a, C, D, E> {
-    app_client: AppClient<'a>,
+struct WebRTCConnection<C, D, E> {
     webrtc_api: WebRtcApi<C, D, E>,
+    sdp: Box<WebRtcSdp>,
+    server: Option<WebRtcGrpcServer<GrpcServer<WebRtcGrpcBody>>>,
+    robot: Arc<Mutex<LocalRobot>>,
 }
-
-impl<'a, C, D, E> WebRtcConnector<'a, C, D, E>
+impl<C, D, E> WebRTCConnection<C, D, E>
 where
     C: Certificate,
     D: DtlsConnector,
-    E: WebRtcExecutor<Pin<Box<dyn Future<Output = ()> + Send>>> + Clone + 'a,
+    E: WebRtcExecutor<Pin<Box<dyn Future<Output = ()> + Send>>> + Clone,
 {
-    async fn open_data_channel(self) -> Result<(Channel, WebRtcApi<C, D, E>), ServerError> {
-        let mut api = {
-            let mut api = self.webrtc_api;
-            let _app = self.app_client;
-            let answer = api
-                .answer()
-                .await
-                .map_err(|e| ServerError::Other(e.into()))?;
-            api.run_ice_until_connected(answer)
-                .timeout(std::time::Duration::from_secs(2))
-                .await
-                .ok_or(ServerError::ServerConnectionTimeout)?
-                .map_err(|e| ServerError::Other(e.into()))?;
-            api
-        };
-        let c = api
+    async fn open_data_channel(&mut self) -> Result<(), ServerError> {
+        self.webrtc_api
+            .run_ice_until_connected(&self.sdp)
+            .timeout(std::time::Duration::from_secs(6))
+            .await
+            .ok_or(ServerError::ServerConnectionTimeout)?
+            .map_err(|e| ServerError::Other(e.into()))?;
+
+        let c = self
+            .webrtc_api
             .open_data_channel()
             .timeout(std::time::Duration::from_secs(6))
             .await
             .ok_or(ServerError::ServerConnectionTimeout)?
             .map_err(|e| ServerError::Other(e.into()))?;
-        Ok((c, api))
+        let srv = WebRtcGrpcServer::new(
+            c,
+            GrpcServer::new(self.robot.clone(), WebRtcGrpcBody::default()),
+        );
+        let _ = self.server.insert(srv);
+        Ok(())
+    }
+    async fn run(&mut self) -> Result<(), ServerError> {
+        if self.server.is_none() {
+            return Err(ServerError::ServerConnectionNotConfigured);
+        }
+        let srv = self.server.as_mut().unwrap();
+        loop {
+            if let Err(e) = srv.next_request().await {
+                return Err(ServerError::Other(Box::new(e)));
+            }
+        }
     }
 }
 
-struct WebRtcListener<'a, C, D, E> {
-    ip: Ipv4Addr,
-    app_client: AppClient<'a>,
-    exec: E,
-    cert: Rc<C>,
-    dtls: D,
+pin_project_lite::pin_project! {
```

Comment:
Can you help me understand this does and why it is needed? Above my current level of understanding.

Reply:
this allow us to worked with a self-referential struct, without is F would have to be Unpin but not all future can fulfill this requirement. 
Using pin_project we can make is the field future is self-referential thus allowing to invoke poll without Unpin.
A good read : https://fasterthanli.me/articles/pin-and-suffering albeit quite long